airbyte-commons/src/main/java/io/airbyte/commons/util/MoreIterators.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.commons.util;
+
+import com.google.common.collect.AbstractIterator;
+import io.airbyte.commons.concurrency.SafeVoidCallable;
+import io.airbyte.commons.concurrency.VoidCallable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class MoreIterators {
+
+  /**
+   * Create an iterator from elements
+   *
+   * @param elements element to put in iterator
+   * @param <T> type
+   * @return iterator with all elements
+   */
+  public static <T> Iterator<T> of(T... elements) {
+    return Arrays.asList(elements).iterator();
+  }
+
+  /**
+   * Create a list from an iterator
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return list
+   */
+  public static <T> List<T> toList(Iterator<T> iterator) {
+    final List<T> list = new ArrayList<>();
+    while (iterator.hasNext()) {
+      list.add(iterator.next());
+    }
+    return list;
+  }
+
+  /**
+   * Create a set from an iterator
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return set
+   */
+  public static <T> Set<T> toSet(Iterator<T> iterator) {
+    final Set<T> set = new HashSet<>();
+    while (iterator.hasNext()) {
+      set.add(iterator.next());
+    }
+    return set;
+  }
+
+  /**
+   * Create an iterator that only executes the given supplier after the first invocation of hasNext.
+   * When fully consumed it will close the input stream.
+   *
+   * @param streamSupplier stream to be supplied lazily
+   * @param <T> type
+   * @return lazy, auto closing iterator
+   */
+  public static <T> Iterator<T> streamSupplierToLazyIterator(Supplier<Stream<T>> streamSupplier) {
```

Comment:
It is a confusing interface: A supplier can supply forever. Can you add a comment to explain that suppler::get will only be called once?

Reply:
you can rename it to: `toLazyIterator`

Reply:
or `lazyIterator`

Reply:
these 2 functions (`streamSupplierToLazyIterator`, `iteratorSupplierToLazyIterator `) cannot both be named `lazyIterator` due to type erasure. open to other less verbose namspacing.

Reply:
Indeed, didn't pay attention to the supplier piece!